### PR TITLE
[STM32] Unify flash drivers

### PR DESCRIPTION
**Pull request by utzig**
This updates the STM32 MCUs to be based on a single flash driver for the most part, leaving just the erase operation in a MCU specific flash driver.
    
Some new syscfgs were added that should be used to provided the configuration:

* `STM32_FLASH_SIZE_KB`: Total size of the flash per BSP.
* `STM32_FLASH_IS_LINEAR`: 1 if all sectors are sized equaly, 0 otherwise.
* `STM32_FLASH_NUM_AREAS`: Number of sectors if this flash is not linear.
* `STM32_FLASH_SECTOR_SIZE`: Size of sectors if the flash is linear.
* `MCU_FLASH_ERASED_VAL`: Value read from erased flash per MCU family.
    
A `const uint32_t stm32_flash_sectors[]` array of sector beginning addresses should be provided by MCUs that have non-linear flash.

### CODE REVIEW

**Comment by utzig** [comment]
I am having a few issues running on F1/F3 but those also happen on `master` so I'll have to check out where it broke. A few `FIXME` are left behind in the PR that I would rather leave there.

This also superseeds #1484

**Comment by utzig** [comment]
Well, fixed some left behind BSPs as well. Btw `STM32F0` was not tested, and fixes that might apply will be added later to the Nucleo-F072 PR.

File: hw/bsp/nucleo-f413zh/src/hal_bsp.c [hunk 1]
```diff
@@ -60,18 +59,9 @@ const uint32_t stm32f4_flash_sectors[] = {
...
-
-const struct hal_flash stm32f4_flash_dev = {
-    .hf_itf = &stm32f4_flash_funcs,
-    .hf_base_addr = 0x08000000,
-    .hf_size = 1536 * 1024,
-    .hf_sector_cnt = NAREAS - 1,
-    .hf_align = 1,
-    .hf_erased_val = 0xff,
-};
+#define SZ (sizeof(stm32_flash_sectors)/sizeof(stm32_flash_sectors[0]))
```
Comment (kasjer) [inline hw/bsp/nucleo-f413zh/src/hal_bsp.c]: other places have spaces around /

File: hw/bsp/nucleo-f746zg/src/hal_bsp.c [hunk 2]
```diff
@@ -80,8 +80,23 @@ static struct stm32_pwm_conf  stm32_pwm_config[PWM_CNT] = {
...
+    0x08018000,     /* 32kB  */
+    0x08020000,     /* 128kB */
+    0x08040000,     /* 256kB */
+    0x08080000,     /* 256kB */
+    0x080c0000,     /* 256kB */
+    0x08100000,     /* End of flash */
+};
 
-#endif
+#define SZ (sizeof(stm32_flash_sectors)/sizeof(stm32_flash_sectors[0]))
```
Comment (kasjer) [inline hw/bsp/nucleo-f746zg/src/hal_bsp.c]: other places have spaces around /

File: hw/bsp/nucleo-f767zi/src/hal_bsp.c [hunk 3]
```diff
@@ -80,8 +80,28 @@ static struct stm32_pwm_conf  stm32_pwm_config[PWM_CNT] = {
...
+    0x080c0000,     /* 256kB */
+    0x08100000,     /* 256kB */
+    0x08140000,     /* 256kB */
+    0x08180000,     /* 256kB */
+    0x081c0000,     /* 256kB */
+    0x08200000,     /* End of flash */
+};
 
-#endif
+#define SZ (sizeof(stm32_flash_sectors)/sizeof(stm32_flash_sectors[0]))
```
Comment (kasjer) [inline hw/bsp/nucleo-f767zi/src/hal_bsp.c]: other places have spaces around /

File: hw/bsp/olimex_stm32-e407_devboard/src/hal_bsp.c [hunk 4]
```diff
@@ -68,18 +67,9 @@ const uint32_t stm32f4_flash_sectors[] = {
...
-
-const struct hal_flash stm32f4_flash_dev = {
-    .hf_itf = &stm32f4_flash_funcs,
-    .hf_base_addr = 0x08000000,
-    .hf_size = 1024 * 1024,
-    .hf_sector_cnt = NAREAS - 1,
-    .hf_align = 1,
-    .hf_erased_val = 0xff,
-};
+#define SZ (sizeof(stm32_flash_sectors)/sizeof(stm32_flash_sectors[0]))
```
Comment (kasjer) [inline hw/bsp/olimex_stm32-e407_devboard/src/hal_bsp.c]: other places have spaces around /

File: hw/bsp/stm32f429discovery/src/hal_bsp.c [hunk 5]
```diff
@@ -69,18 +67,9 @@ const uint32_t stm32f4_flash_sectors[] = {
...
-
-const struct hal_flash stm32f4_flash_dev = {
-    .hf_itf = &stm32f4_flash_funcs,
-    .hf_base_addr = 0x08000000,
-    .hf_size = 2048 * 1024,
-    .hf_sector_cnt = NAREAS - 1,
-    .hf_align = 1,
-    .hf_erased_val = 0xff,
-};
+#define SZ (sizeof(stm32_flash_sectors)/sizeof(stm32_flash_sectors[0]))
```
Comment (kasjer) [inline hw/bsp/stm32f429discovery/src/hal_bsp.c]: other places have spaces around /

**Pull request merged**